### PR TITLE
🎉Source Klaviyo: migrate to low-code

### PR DESCRIPTION
**Pull request by askarpets**
## What

Resolves https://github.com/airbytehq/airbyte-internal-issues/issues/6657

## How

Migrated streams `profiles`, `global_exclusions`, `events`, `email_templates`, `metrics`, and `lists` to low-code

## Recommended reading order

1. `source_klaviyo/manifest.yaml`
2. `source_klaviyo/source.py`
3. `source_klaviyo/components/datetime_based_cursor.py`

## 🚨 User Impact 🚨
No

## Pre-merge Actions

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- Unit & integration tests added

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- Create a non-forked branch based on this PR and test the below items on it
- Build is successful
- If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).

</details>

### CODE REVIEW

File: airbyte-integrations/connectors/source-klaviyo/integration_tests/abnormal_state.json [hunk 1]
```diff
@@ -2,67 +2,95 @@
...
+        "name": "lists"
+      }
     }
   },
   {
     "type": "STREAM",
     "stream": {
       "stream_state": {
-        "updated_at": "2120-10-10 00:00:00",
+        "updated_at": "2120-10-10T00:00:00+00:00",
```
Comment (lazebnyi) [inline airbyte-integrations/connectors/source-klaviyo/integration_tests/abnormal_state.json:64]: Why we need update state format?

File: airbyte-integrations/connectors/source-klaviyo/source_klaviyo/streams.py [hunk 2]
```diff
@@ -99,13 +101,13 @@ def get_updated_state(self, current_stream_state: MutableMapping[str, Any], late
...
 
-    def backoff_time(self, response: requests.Response) -> Optional[float]:
+    def backoff_time(self, response: Response) -> Optional[float]:
         if response.status_code == 429:
             retry_after = response.headers.get("Retry-After")
             retry_after = float(retry_after) if retry_after else None
             if retry_after and retry_after >= self.max_time:
                 raise KlaviyoBackoffError(
-                    f"Stream {self.name} has reached rate limit with 'Retry-After' of {retry_after} seconds, exit from stream."
+                    f"Stream {self.name} has reached rate limit with 'Retry-After' of {retry_after} seconds, " "exit from stream."
```
Comment (lazebnyi) [inline airbyte-integrations/connectors/source-klaviyo/source_klaviyo/streams.py:110]: why do we need brackets before `exit`?

File: airbyte-integrations/connectors/source-klaviyo/README.md [hunk 3]
```diff
@@ -30,7 +30,7 @@ See `sample_files/sample_config.json` for a sample config file.
 poetry run source-klaviyo spec
 poetry run source-klaviyo check --config secrets/config.json
 poetry run source-klaviyo discover --config secrets/config.json
-poetry run source-klaviyo read --config secrets/config.json --catalog sample_files/configured_catalog.json
+poetry run source-klaviyo read --config secrets/config.json --catalog integration_tests/configured_catalog.json
```
Comment (lazebnyi) [inline airbyte-integrations/connectors/source-klaviyo/README.md:33]: Have you run any regression tests on it?

**Review by lazebnyi — CHANGES_REQUESTED** [review]

File: airbyte-integrations/connectors/source-klaviyo/integration_tests/abnormal_state.json (hunk 1 above)
Comment (askarpets) [reply airbyte-integrations/connectors/source-klaviyo/integration_tests/abnormal_state.json:64]: I updated these so the formats match what is returned from the API

File: airbyte-integrations/connectors/source-klaviyo/source_klaviyo/streams.py (hunk 2 above)
Comment (askarpets) [reply airbyte-integrations/connectors/source-klaviyo/source_klaviyo/streams.py]: Removed, thanks for noticing that!

File: airbyte-integrations/connectors/source-klaviyo/README.md (hunk 3 above)
Comment (askarpets) [reply airbyte-integrations/connectors/source-klaviyo/README.md:33]: I've compared the records counts/structure and API calls made by the connector manually. Should I use the new tool for that?

**Review by alafanechere — CHANGES_REQUESTED** [review]
@askarpets I ran your new connector version by the regression tool. 
Here are my findings.
### Record count
They are correct. No missing records.

### Checkpointing
The new version produces far less state messages than the previous one. 
I believe a single state message is produced per stream on the new version.
Do you have an explanation for this change in checkpointing behavior? Is it alarming?

### Breaking schema changes

#### `campaigns` stream
The following fields were removed. Why?
```
attributes.channel
attributes.message 
```

#### `profiles` and `global_exclusions` streams
Some fields were renamed. Can we keep the same field naming please to avoid a breaking change on these streams:

```
attributes.subscriptions.email.timestamp > consent_timestamp
attributes.subscriptions.email.suppressions > suppression
attributes.subscriptions.sms.timestamp > consent_timestamp
```

### `events` stream
The `attributes.event_properties.$attribution `was removed.

## Conclusion
I am requesting changes because:
### A. I think we can limit the scope of breaking changes.

I need to understand if there's a blocker forcing you to make these schema changes.
If they are mandatory we should use [`scopedImpact` in the breaking changes definition](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connectors/source-facebook-marketing/metadata.yaml#L36) to not bother users who are not syncing the affected streams.

I would suggest making the API upgrade in a different and follow up PR to reduce the number of moving parts in this migration.

### B. I'd like to make sure the checkpointing is correct
Please share why we have less state messages and if it's safe or not.

**Comment by askarpets** [comment]
Hi @alafanechere,
please find my responses below:

**checkpointing:** this feature has been intentionally disabled for low-code streams - please see [declarative_stream.py](https://github.com/airbytehq/airbyte/blob/master/airbyte-cdk/python/airbyte_cdk/sources/declarative/declarative_stream.py#L145).

**`campaigns` stream:** these fields are not present in the response schema for API version `2024-02-15`, however, this info can be found in `campaign_messages.attributes` section of the new `campaigns_detailed` stream.
I had to upgrade the API version in order to be able to fetch additional data without making extra requests according to this [ticket](https://github.com/airbytehq/product-request-backlog/issues/45).

**`profiles`, `global_exclusions` streams:** these fields have been renamed in the new API version. Should I add transformations to bring previous names back?

**Review by alafanechere — CHANGES_REQUESTED** [review]
Thanks @askarpets for your reply!
🆗 for checkpointing, I've no concern as long as we understand what's happening.

I'm re-requesting changes because I really think https://github.com/airbytehq/product-request-backlog/issues/45 should be treated in a followup PR to minimize the changes and make this migration non-breaking. Can you consider not making the API upgrade in this PR 🙏 ?

**Comment by askarpets** [comment]
@alafanechere I have reverted all breaking changes, so now it is a pure migration. Please review and let me know if you see any issues. Thanks!

**Review by alafanechere — CHANGES_REQUESTED** [review]
@askarpets I really appreciate the change.
The regression test tool still shows me a light discrepancy in the `campaigns` stream schema:
**Breaking**:
`attributes.channel` and `attributes.message` fields got removed.
Is it expected?
**Non-breaking**:
`relationships.campaign-messages` field was added.

**Review by alafanechere — APPROVED** [review]
Ignore my last review, I did not rebuild the image...
Now the test report looks great.
* Schemas are matching 
* Record counts are matching

File: airbyte-integrations/connectors/source-klaviyo/metadata.yaml [hunk 4]
```diff
@@ -33,16 +33,16 @@ data:
...
     breakingChanges:
-      1.0.0:
-        message: In this release, for 'events' stream changed type of 'event_properties/items/quantity' field from integer to number. Users will need to refresh the source schema and reset events streams after upgrading.
-        upgradeDeadline: "2023-11-30"
       2.0.0:
         message: In this release, streams 'campaigns', 'email_templates', 'events', 'flows', 'global_exclusions', 'lists', and 'metrics' are now pulling data using latest API which has a different schema. Users will need to refresh the source schemas and reset these streams after upgrading.
         upgradeDeadline: "2023-11-30"
+      1.0.0:
+        message: In this release, for 'events' stream changed type of 'event_properties/items/quantity' field from integer to number. Users will need to refresh the source schema and reset events streams after upgrading.
+        upgradeDeadline: "2023-11-30"
```
[the remark this answers is not in the document]
Comment (alafanechere) [reply airbyte-integrations/connectors/source-klaviyo/metadata.yaml:41]: Can you remove it?

[the remark this answers is not in the document]
Comment (askarpets) [reply airbyte-integrations/connectors/source-klaviyo/metadata.yaml:41]: This was present here before, I just changed the order to make sure the latest breaking change description is at the top level

[the remark this answers is not in the document]
Comment (alafanechere) [reply airbyte-integrations/connectors/source-klaviyo/metadata.yaml:41]: Sorry I'm dumb today 😄

**Review by lazebnyi — APPROVED** [review]
Lgtm

**Pull request merged**